File: backend/app/providers/gpt.py
"""Provedor GPT: OAuth do Codex CLI (ChatGPT) + contas extras coladas no painel."""

from __future__ import annotations

from typing import Any

from app.formatting import iso_or_none, pick
from app.http_util import http_json
from app.local.gpt_oauth import gpt_token_candidates, gpt_token_expired
from app.store import provider as provider_cfg

GPT_USAGE_URL = "https://chatgpt.com/backend-api/wham/usage"
GPT_USER_AGENT = "codex-cli"
SESSION_MAX_S = 8 * 3600


def gpt_fail(msg: str) -> dict[str, Any]:
    return {
        "ok": False,
        "error": msg,
        "session_percent": None,
        "session_resets_at": None,
        "weekly_percent": None,
        "weekly_resets_at": None,
        "plan": None,
    }


def _used_percent(value: Any) -> float | None:
    """WHAM devolve 0–100. Não tratar 0.5 como 50%."""
    if value is None:
        return None
    try:
        n = float(value)
    except (TypeError, ValueError):
        return None
    if n < 0:
        n = 0.0
    if n > 100:
        n = 100.0
    return round(n, 1)


def _extract_window(obj: Any) -> tuple[float | None, str | None, int | None]:
    if not isinstance(obj, dict) or not obj:
        return None, None, None
    pct = _used_percent(pick(obj.get("used_percent"), obj.get("utilization"), obj.get("percent")))
    reset = iso_or_none(pick(obj.get("reset_at"), obj.get("resets_at"), obj.get("resetsAt")))
    secs_raw = pick(obj.get("limit_window_seconds"), obj.get("window_seconds"))
    secs: int | None = None
    if secs_raw is not None:
        try:
            secs = int(secs_raw)
        except (TypeError, ValueError):
            secs = None
    return pct, reset, secs


def _is_session(secs: int | None) -> bool:
    return secs is not None and 0 < secs <= SESSION_MAX_S


def parse_gpt_payload(data: dict[str, Any]) -> dict[str, Any]:
    rl = data.get("rate_limit") if isinstance(data.get("rate_limit"), dict) else data
    primary = pick(
        rl.get("primary_window") if isinstance(rl, dict) else None,
        data.get("five_hour"),
        data.get("fiveHour"),
        data.get("five_hour_limit"),
    )
    secondary = pick(
        rl.get("secondary_window") if isinstance(rl, dict) else None,
        data.get("weekly"),
        data.get("weekly_limit"),
        data.get("seven_day"),
    )

    windows: list[tuple[float | None, str | None, int | None]] = []
    for raw in (primary, secondary):
        pct, reset, secs = _extract_window(raw)
        if pct is None and reset is None and secs is None:
            continue
        windows.append((pct, reset, secs))

    session_pct: float | None = None
    session_reset: str | None = None
    weekly_pct: float | None = None
    weekly_reset: str | None = None

    for pct, reset, secs in windows:
        if session_pct is None and (secs is None or _is_session(secs)):
            session_pct, session_reset = pct, reset
            continue
        if weekly_pct is None:
            weekly_pct, weekly_reset = pct, reset
        elif session_pct is None:
            session_pct, session_reset = pct, reset

    plan = data.get("plan_type")
    plan_s = str(plan).strip() if plan else None
    if plan_s == "":
        plan_s = None

    ok = session_pct is not None or weekly_pct is not None
    return {
        "ok": ok,
        "error": None if ok else "resposta GPT sem janelas de cota",
        "session_percent": session_pct,
        "session_resets_at": session_reset,
        "weekly_percent": weekly_pct,
        "weekly_resets_at": weekly_reset,
        "plan": plan_s,
    }


def fetch_gpt_one(token: str, account_id: str | None = None) -> dict[str, Any]:
    token = (token or "").strip()
    if not token:
        return gpt_fail("sem token GPT")
    headers = {
        "Authorization": f"Bearer {token}",
        "Accept": "application/json",
        "User-Agent": GPT_USER_AGENT,
    }
    if account_id:
        headers["ChatGPT-Account-Id"] = account_id.strip()
    try:
        data = http_json(GPT_USAGE_URL, headers=headers, provider="GPT")
    except RuntimeError as exc:
        return gpt_fail(str(exc))
    if not isinstance(data, dict):
        return gpt_fail("resposta GPT inesperada")
    parsed = parse_gpt_payload(data)
    if parsed.get("ok"):
        return parsed
    return gpt_fail(str(parsed.get("error") or "resposta GPT sem janelas de cota"))


def _fetch_gpt_local(cands: list[tuple[str, str, str | None, int | None]]) -> dict[str, Any]:
    last_err = "sem token GPT"
    for _source, token, account_id, _exp in cands:
        if gpt_token_expired(token):
            last_err = "OAuth expirado; abra o Codex neste computador (`codex login`)"
            continue
        result = fetch_gpt_one(token, account_id)
        if result.get("ok"):
            return result
        last_err = str(result.get("error") or last_err)
    return gpt_fail(last_err)


def fetch_gpt_accounts(cfg: dict) -> list[dict[str, Any]]:
    accounts: list[dict[str, Any]] = []
    p = provider_cfg(cfg, "gpt")

    if not p.get("hidden"):
        local_cands = gpt_token_candidates(cfg)
        if local_cands:
            result = _fetch_gpt_local(local_cands)
            accounts.append({"id": "local", "label": str(p.get("local_label") or ""), **result})

    extra = list(p.get("accounts") or [])
    if not extra:
        legacy = str(p.get("paste_secret") or "").strip()
        if legacy:
            extra = [{"id": "legacy", "label": "", "secret": legacy}]
    for acc in extra:
        token = str(acc.get("secret") or "").strip()
        label = str(acc.get("label") or "").strip()
        aid = str(acc.get("id") or "extra")
        result = fetch_gpt_one(token)
        accounts.append({"id": aid, "label": label, **result})
    return accounts

File: backend/src/providers/gpt.ts
/**
 * Provedor GPT: OAuth do Codex CLI (ChatGPT) + contas extras coladas no painel.
 * Port of backend-python-legacy/app/providers/gpt.py
 */
import { isoOrNone, pick } from "../formatting.js";
import { httpJson } from "../httpClient.js";
import { gptTokenCandidates, gptTokenExpired } from "../local/gptOauth.js";
import { provider as providerCfg } from "../store.js";

export const GPT_USAGE_URL = "https://chatgpt.com/backend-api/wham/usage";
export const GPT_USER_AGENT = "codex-cli";
export const SESSION_MAX_S = 8 * 3600;

export function gptFail(msg: string): Record<string, unknown> {
  return {
    ok: false,
    error: msg,
    session_percent: null,
    session_resets_at: null,
    weekly_percent: null,
    weekly_resets_at: null,
    plan: null,
  };
}

function usedPercent(value: unknown): number | null {
  if (value === null || value === undefined) return null;
  let n: number;
  try {
    n = Number(value);
    if (Number.isNaN(n) || !Number.isFinite(n)) return null;
  } catch {
    return null;
  }
  if (n < 0) n = 0.0;
  if (n > 100) n = 100.0;
  return Math.round(n * 10) / 10;
}

function extractWindow(obj: unknown): [number | null, string | null, number | null] {
  if (obj === null || typeof obj !== "object" || Array.isArray(obj) || Object.keys(obj as object).length === 0) return [null, null, null];
  const d = obj as Record<string, unknown>;
  const pct = usedPercent(pick(d.used_percent, d.utilization, d.percent));
  const reset = isoOrNone(pick(d.reset_at, d.resets_at, d.resetsAt));
  const secsRaw = pick(d.limit_window_seconds, d.window_seconds);
  let secs: number | null = null;
  if (secsRaw !== null && secsRaw !== undefined) {
    try {
      secs = Math.trunc(Number(secsRaw));
      if (Number.isNaN(secs)) secs = null;
    } catch {
      secs = null;
    }
    // Python: int(secs_raw) — if it fails sets None
    if (secs !== null && Number.isNaN(secs)) secs = null;
  }
  return [pct, reset, secs];
}

function isSession(secs: number | null): boolean {
  return secs !== null && secs !== undefined && 0 < secs && secs <= SESSION_MAX_S;
}

export function parseGptPayload(data: Record<string, unknown>): Record<string, unknown> {
  const rl = data.rate_limit !== null && typeof data.rate_limit === "object" && !Array.isArray(data.rate_limit)
    ? (data.rate_limit as Record<string, unknown>)
    : (data as Record<string, unknown>);
  const primary = pick(
    rl !== data ? (rl as Record<string, unknown>).primary_window : null,
    data.five_hour,
    data.fiveHour,
    data.five_hour_limit,
  );
  const secondary = pick(
    rl !== data ? (rl as Record<string, unknown>).secondary_window : null,
    data.weekly,
    data.weekly_limit,
    data.seven_day,
  );

  const windows: Array<[number | null, string | null, number | null]> = [];
  for (const raw of [primary, secondary]) {
    const [pct, reset, secs] = extractWindow(raw);
    if (pct === null && reset === null && secs === null) continue;
    windows.push([pct, reset, secs]);
  }

  let sessionPct: number | null = null;
  let sessionReset: string | null = null;
  let weeklyPct: number | null = null;
  let weeklyReset: string | null = null;

  // Preserve exact assignment order from Python:
  for (const [pct, reset, secs] of windows) {
    if (sessionPct === null && (secs === null || isSession(secs))) {
      sessionPct = pct;
      sessionReset = reset;
      continue;
    }
    if (weeklyPct === null) {
      weeklyPct = pct;
      weeklyReset = reset;
    } else if (sessionPct === null) {
      sessionPct = pct;
      sessionReset = reset;
    }
  }

  const planRaw = data.plan_type;
  let planS: string | null = planRaw ? String(planRaw).trim() : null;
  if (planS === "") planS = null;
  if (!planS) planS = null;

  const ok = sessionPct !== null || weeklyPct !== null;
  return {
    ok,
    error: ok ? null : "resposta GPT sem janelas de cota",
    session_percent: sessionPct,
    session_resets_at: sessionReset,
    weekly_percent: weeklyPct,
    weekly_resets_at: weeklyReset,
    plan: planS,
  };
}

export async function fetchGptOne(token: string, accountId?: string | null): Promise<Record<string, unknown>> {
  token = (token ?? "").trim();
  if (!token) return gptFail("sem token GPT");
  const headers: Record<string, string> = {
    Authorization: `Bearer ${token}`,
    Accept: "application/json",
    "User-Agent": GPT_USER_AGENT,
  };
  if (accountId) headers["ChatGPT-Account-Id"] = accountId.trim();
  let data: unknown;
  try {
    data = await httpJson(GPT_USAGE_URL, { headers, provider: "GPT" });
  } catch (e) {
    return gptFail(String(e));
  }
  if (data === null || typeof data !== "object" || Array.isArray(data)) return gptFail("resposta GPT inesperada");
  const parsed = parseGptPayload(data as Record<string, unknown>);
  if (parsed.ok) return parsed;
  return gptFail(String(parsed.error ?? "resposta GPT sem janelas de cota"));
}

async function fetchGptLocal(
  cands: Array<[string, string, string | null, number | null]>,
): Promise<Record<string, unknown>> {
  let lastErr = "sem token GPT";
  for (const [, token, accountId] of cands) {
    if (gptTokenExpired(token)) {
      lastErr = "OAuth expirado; abra o Codex neste computador (`codex login`)";
      continue;
    }
    const result = await fetchGptOne(token, accountId);
    if (result.ok) return result;
    lastErr = String(result.error ?? lastErr);
  }
  return gptFail(lastErr);
}

export async function fetchGptAccounts(cfg: Record<string, unknown>): Promise<Array<Record<string, unknown>>> {
  const accounts: Array<Record<string, unknown>> = [];
  const p = providerCfg(cfg, "gpt") as Record<string, unknown>;

  if (!p.hidden) {
    const localCands = gptTokenCandidates(cfg);
    if (localCands.length > 0) {
      const result = await fetchGptLocal(localCands);
      accounts.push({ id: "local", label: String(p.local_label ?? ""), ...result });
    }
  }

  let extra = Array.isArray(p.accounts) ? [...(p.accounts as unknown[])] : [];
  if (extra.length === 0) {
    const legacy = String(p.paste_secret ?? "").trim();
    if (legacy) extra = [{ id: "legacy", label: "", secret: legacy }];
  }
  for (const accRaw of extra) {
    const acc = accRaw as Record<string, unknown>;
    const token = String(acc.secret ?? "").trim();
    const label = String(acc.label ?? "").trim();
    const aid = String(acc.id ?? "extra");
    const result = await fetchGptOne(token);
    accounts.push({ id: aid, label, ...result });
  }
  return accounts;
}

export const gpt_fail = gptFail;
export const parse_gpt_payload = parseGptPayload;
export const fetch_gpt_one = fetchGptOne;
export const fetch_gpt_accounts = fetchGptAccounts;
